### PR TITLE
Use docker login when possible

### PR DESCRIPTION
**Pull request by rcoh**
## Motivation and Context

We've had frequent throttling errors from docker

## Description
This PR adds an optional job to acquire credentials from docker. the credentials are saved into output and encrypted with a repo secret. The credential saved is only usable for ECR and is timescoped to 12 hours.

## Testing
- this CI job
- unit tests

## Checklist

- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: .github/scripts/acquire-build-image [hunk 1]
```diff
@@ -41,7 +42,8 @@ class Platform(Enum):
 
 # Script context
 class Context:
-    def __init__(self, start_path, script_path, tools_path, user_id, image_tag, allow_local_build, github_actions):
+    def __init__(self, start_path, script_path, tools_path, user_id, image_tag, allow_local_build, github_actions,
+                 encrpyted_docker_password, docker_passphrase):
```
Comment (jdisanti) [inline .github/scripts/acquire-build-image:46]:
```suggestion
                 encrypted_docker_password, docker_passphrase):
```

File: .github/scripts/acquire-build-image [hunk 2]
```diff
@@ -50,6 +52,8 @@ class Context:
         self.image_tag = image_tag
         self.allow_local_build = allow_local_build
         self.github_actions = github_actions
+        self.encrypted_docker_password = encrpyted_docker_password
```
Comment (jdisanti) [inline .github/scripts/acquire-build-image:55]:
```suggestion
        self.encrypted_docker_password = encrypted_docker_password
```

File: .github/scripts/acquire-build-image [hunk 3]
```diff
@@ -101,17 +106,19 @@ class Shell:
         print(stderr)
         print("-------------------")
 
-        not_found_messages = ["not found: manifest unknown"]
-        throttle_messages = ["toomanyrequests: Rate exceeded", "toomanyrequests: Data limit exceeded"]
```
Comment (jdisanti) [inline .github/scripts/acquire-build-image:105]: Looks like this was unintentionally reverted. We definitely want this "Data limit exceeded" retry for forked PRs.

File: .github/scripts/acquire-build-image [hunk 4]
```diff
@@ -211,15 +240,18 @@ def acquire_build_image(context=Context.default(), shell=Shell()):
...
-    def test_context(self, allow_local_build=False, github_actions=False):
+    def test_context(self, github_actions=False, allow_local_build=False, encrypted_docker_password=None,
+                     docker_passphrase=None):
         return Context(
             start_path="/tmp/test/start-path",
             script_path="/tmp/test/script-path",
             tools_path="/tmp/test/tools-path",
             user_id="123",
             image_tag="someimagetag",
+            encrpyted_docker_password=encrypted_docker_password,
```
Comment (jdisanti) [inline .github/scripts/acquire-build-image:251]:
```suggestion
            encrypted_docker_password=encrypted_docker_password,
```

File: .github/scripts/docker-image-hash [hunk 5]
```diff
@@ -11,4 +11,4 @@ set -eo pipefail
 cd "$(dirname "$0")"
 cd "$(git rev-parse --show-toplevel)"
 
-git ls-files -s --full-name "tools" | git hash-object --stdin
+git ls-files -s --full-name "tools" | grep -v acquire-build-image | git hash-object --stdin
```
Comment (jdisanti) [inline .github/scripts/docker-image-hash:14]: This change isn't needed anymore. The script was moved out of tools.

File: .github/scripts/acquire-build-image (hunk 3 above)
Comment (rcoh) [reply .github/scripts/acquire-build-image:105]: oh thanks I saw that and TODO but forgot

**Review by jdisanti — APPROVED** [review]

**Review by ysaito1001 — APPROVED** [review]

**Pull request merged**